### PR TITLE
Add codeql-analysis.yml GitHub Actions Workflow

### PR DESCRIPTION
**Pull request by JLLeitschuh**
Hello! Your friendly neighborhood security researcher here!

GitHub is slowly (not entirely there yet) migrating away from LGTM.com
to establishing a way for security researchers to work with CodeQL databases
fully through GitHub.com. This 'new' way requires that databases get built
via GitHub actions and uploaded to GitHub.

The old way, supported by the `.lgtm.yaml` file will continue to exist for a while
(and I'll continue to use it), but as a researcher, if the Jetty project also built a
CodeQL database and uploaded it, I'd find it incredibly useful.

In summary, this PR adds a GitHub action to build a CodeQL database and runs
GitHub's CodeQL analysis tool which audits the code for security vulnerabilities.

### CODE REVIEW

File: .github/workflows/codeql-analysis.yml
Comment (joakime) [inline .github/workflows/codeql-analysis.yml]: Just hardcode `jetty-10.0.x`, `jetty-11.0.x`, and `jetty-12.0.x`
Note: Jetty 9.x is being deprecated soon. See: #7958

File: .github/workflows/codeql-analysis.yml
Comment (joakime) [inline .github/workflows/codeql-analysis.yml:43]: Jetty 12 will requires Java 17.

File: .github/workflows/codeql-analysis.yml
Comment (JLLeitschuh) [reply .github/workflows/codeql-analysis.yml]: If you support branching names like `9.4.x`, and will use that pattern for future development, wouldn't it be this?
`[ 'jetty-10.[1-9]?[0-9].x', 'jetty-11.[1-9]?[0-9].x', 'jetty-12.[1-9]?[0-9].x' ]`

File: .github/workflows/codeql-analysis.yml
Comment (JLLeitschuh) [reply .github/workflows/codeql-analysis.yml:43]: Thanks fixed! 🙂

**Comment by JLLeitschuh** [comment]
Once this is merged, the maintainers of the eclipse jetty project will be able to see the alerts here.
https://github.com/eclipse/jetty.project/security/code-scanning

Also, merging this will establish a new baseline, and any PRs that unintentionally introduce a new security vulnerability will be flagged 😄

**Comment by JLLeitschuh** [comment]
I think this PR is in a complete state. The one check 'Code scanning results / CodeQL' will not pass this time, but for other PRs it will pass because it will compare against what's on the default branch as a baseline.

**Comment by joakime** [comment]
This is interesting, but we'll need some time to absorb what this means to our processes.

**Comment by JLLeitschuh** [comment]
If you don't want to run this job on PRs in general, we can always remove these lines:

https://github.com/eclipse/jetty.project/pull/7961/files#diff-63bd641104d10e25f141d518a16b22a151d125e12701df2f9e79734b23b90188R6-R8

File: .github/workflows/codeql-analysis.yml
Comment (olamy) [inline .github/workflows/codeql-analysis.yml:24]: we do not have any javascript here. So there is no real need of a matrix

File: .github/workflows/codeql-analysis.yml
Comment (olamy) [inline .github/workflows/codeql-analysis.yml:8]: not sure we want to run this on PRs.
just a weekly on hardcoded branches might be enough.

File: .github/workflows/codeql-analysis.yml
Comment (joakime) [reply .github/workflows/codeql-analysis.yml:24]: We do have javascript, in the demo apps. Check the bottom of the "files changed" tab in this PR to see where it's finding bugs.

File: .github/workflows/codeql-analysis.yml
Comment (joakime) [reply .github/workflows/codeql-analysis.yml:8]: I'm ok with PRs, it will catch accidentally introduced bugs.

File: .github/workflows/codeql-analysis.yml
Comment (olamy) [reply .github/workflows/codeql-analysis.yml:24]: ah yup demos 🤦

**Comment by olamy** [comment]
> This is interesting, but we'll need some time to absorb what this means to our processes.

correct because we already some reports we do nothing about.
so here we will have to mark the reported errors or they will be sitting here and we will do nothing and this will be YASRWDC (Yet Another Static Report We Don't Care :) )

**Comment by JLLeitschuh** [comment]
> so here we will have to mark the reported errors or they will be sitting here and we will do nothing and this will be YASRWDC (Yet Another Static Report We Don't Care :) )

Not a bad thing, but yes. I'm currently discussing with the GitHub team if there's a way to build the database without reporting query results. They say that there is, but it's a bit of a hack.

**Review by joakime — APPROVED** [review]

**Comment by olamy** [comment]
is it possible to exclude `src/test` from the scanning?

**Comment by JLLeitschuh** [comment]
It is, but I don't remember how to off hand. I suggest consulting the documentation

**Pull request merged**
